### DISCUSSION TITLE
ru/tasks/pose/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9820] [Docs]
# ru/tasks/pose/

Узнай, как использовать Ultralytics YOLOv8  для задач оценки позы. Найди предварительно обученные модели, научись обучать, проверять, предсказывать и экспортировать свои собственные.

https://docs.ultralytics.com/ru/tasks/pose/

**Comment by mihkriss** [comment]
Hello, I want to ask you for help
My task is to determine in the video whether there is a fight or not, using a model based on key points,
Could you tell me how to implement this using yolov8-pose

**Reply by pderrenger** [reply]
Hello! For detecting fights using keypoint analysis with YOLOv8-pose, you can start by utilizing the pretrained pose estimation models such as `yolov8n-pose.pt`. Here's a quick guide on how to set it up for your specific task:

1. **Load Your Model:**
   ```python
   from ultralytics import YOLO
   model = YOLO('yolov8n-pose.pt')
   ```

2. **Run Prediction on Video:**
   You'll need to process your video frame by frame and use the model to predict keypoints. For each frame:
   ```python
   results = model.track('path_to_your_video.mp4', show=True)
   ```
   This will display the video with keypoints detected.

3. **Analyze Keypoints:**
   After detecting keypoints, analyze the keypoint coordinates to determine fight-like movements. This part requires defining your own criteria on what constitutes a fight based on the relationships between keypoints (like rapid changes in distance between certain keypoints).

4. **Loop Through Video:**
   You'll loop through the video, apply step 2, and based on your criteria in step 3, determine if a fight is occurring.

Remember, the effectiveness of your fight detection model largely depends on the criteria you set for identifying fight behavior based on keypoints. It might involve some trial and error to get it right.

Feel free to experiment and adjust the pipeline according to your findings. Good luck with your project! 🚀

**Reply by mihkriss** [reply]
It's a little unclear how to extract key points from a video?

**Reply by pderrenger** [reply]
Hey there! To extract keypoints from a video using the YOLOv8-pose model, you can simply use the following Python code snippet:

```python
from ultralytics import YOLO

# Load your pretrained YOLOv8-pose model
model = YOLO('yolov8n-pose.pt')  

# Run prediction on your video
results = model.track('path/to/your/video.mp4')

# To access the keypoints from the results
for result in results:
    keypoints = result.keypoints  # This gives you the keypoints for each frame
    print(keypoints)
```

This way, you analyze each frame of your video for pose estimation. The `keypoints` will contain information on the detected keypoints in each frame.

Feel free to tweak this example to suit your specific needs. If you have any further questions, just let me know! 😊

**Reply by mihkriss** [reply]
There's a problem
I extracted the coordinates with the fight from data and converted these coordinates of the points into .csv format, how can I further train the model so that it recognizes where there is a fight and where there is not?
Or do you need a different format for storing coordinates?

**Reply by pderrenger** [reply]
Hey there! For training a model to recognize activities like fighting using keypoints from videos, converting the keypoints to CSV is a great first step! Next, you'll need to design or adapt a model that can learn from these coordinates to distinguish between fighting and non-fighting sequences. 🤔

While YOLOv8-pose is excellent for keypoint detection, identifying activities usually requires additional analysis or model architecture, like a time-series analysis model that considers the keypoints' movement over time. You might consider using LSTM (Long Short-Term Memory) networks or 3D ConvNets that can process sequential data alongside your extracted keypoints for recognizing specific actions.

Regarding the format, storing coordinates in .csv is okay for initial data handling, but for training, you might need to convert them into tensors or a format suitable for your specific learning model. The key aspect is to feed the data into your model in a way that preserves the temporal relationships between frames for it to learn the patterns indicating a fight. 

You can start by labeling your data (fight or no fight) and then training a custom model on these sequences. Here’s a simple pseudo-approach in code to get you started:

```python
import torch
import pandas as pd
from torch.utils.data import Dataset, DataLoader

# Load your CSV data
class FightDataset(Dataset):
    def __init__(self, csv_file):
        self.dataframe = pd.read_csv(csv_file)
    
    def __len__(self):
        return len(self.dataframe)
    
    def __getitem__(self, idx):
        keypoints = self.dataframe.iloc[idx, :-1].values  # assuming the last column is your label
        label = self.dataframe.iloc[idx, -1]
        sample = {'keypoints': keypoints, 'label': label}
        return sample

# Example usage
dataset = FightDataset(csv_file='your_data.csv')
dataloader = DataLoader(dataset, batch_size=4, shuffle=True)
```

This is quite a simplified example and may need adjustments based on how your data is structured. Additionally, integrating keypoint time-series data into an activity recognition model is complex and requires custom architecture tailored to your specific task. Keep experimenting, and good luck! 🚀

**Reply by mihkriss** [reply]
Could you please tell me, if I still store coordinates in yaml format, what form it should look like so that I can train it right away like this:
results = model.train(data='custom-pose.yaml', epochs=100, imgsz=640)

**Reply by pderrenger** [reply]
Hey! For training a model with your custom keypoints stored in YAML format, your `custom-pose.yaml` should ideally follow the structure used for datasets in YOLOv8, which generally includes paths to your data, class names, and potentially, specific keypoint annotations format if it diverges from the standard COCO format. Since you're doing keypoint detection, ensuring that your keypoints are correctly labeled and structured is crucial. Here's a simplified example structure for your YAML:

```yaml
# custom-pose.yaml
train: path/to/train/images  # Directory with your training images
val: path/to/val/images      # Directory with your validation images

# Define your classes, e.g., person
nc: 1  
names: ['person']  

# Optionally, specific keypoint names if applicable
keypoints: ['nose', 'left_eye', 'right_eye', ...]
```

For training, the keypoints in your dataset should be annotated in a manner similar to how COCO formats keypoint annotations, typically as an array of `x, y, visibility` values for each keypoint. Your data preparation script should transform these annotations into the required format, pairing each image with its corresponding keypoint annotations, and ensuring the model training process can effectively learn from these patterns.

If training directly on keypoints in a non-standard format, consider writing a custom data loading script that converts your YAML structured keypoints into the necessary format for model training. 

Remember, the key here is to ensure that your YAML file correctly points to your data and annotations in a way that YOLOv8 can understand and effectively learn from! 🚀

**Reply by mihkriss** [reply]
In what format should these points be stored?

**Reply by pderrenger** [reply]
Hey there! To store keypoints for training a pose estimation model like `yolov8n-pose`, the format should include the image paths and corresponding keypoint coordinates (and visibility flags if available). Here's a quick example structure that you might use in your YAML file:

```yaml
# Example structure for your custom-pose.yaml
train: path/to/your/train/images/
val: path/to/your/val/images/

# Assuming a single class (e.g., person)
nc: 1  
names: ['person']

# Define keypoints names (if applicable to your dataset)
keypoints: ['nose', 'left_eye', 'right_eye', ...]
```

And for the annotations, you can align them with existing standards (like COCO), where each keypoint is represented by a list, `[x1, y1, v1, x2, y2, v2, ..., xn, yn, vn]` per image/instance, accompanied by an image ID. The annotations might be stored separately or included in a structured format within your dataset directory. When you train, ensure the annotations are correctly parsed and matched with each training image.

If you've got your keypoints and annotations ready, to train your model, here's how you might proceed:

```python
from ultralytics import YOLO

# Load and train the model
model = YOLO('yolov8n-pose.yaml')  # Ensure this YAML correctly references your custom dataset
results = model.train(data='custom-pose.yaml', epochs=100, imgsz=640)
```

Happy coding, and if you have any more questions, feel free to ask! 😊

**Reply by mihkriss** [reply]
You are talking about anatations, but where should they be stored?
I understand it's in .txt format?
Is it possible to see an example of how the data for processing poses should look there?

**Reply by pderrenger** [reply]
Hello! 🙌 Yes, the annotations for keypoints are typically stored in text (.txt) format alongside your images in the dataset. Each text file corresponds to an image and contains the annotations for that image in a specific format. Here's a quick example of how keypoints might be represented in a .txt file for pose estimation tasks:

```
0 0.51 0.35 0.02 0.02 1 0.55 0.38 0.02 0.02 1 ... 0.45 0.60 0.02 0.02 1
```
Let's break it down:
- Each line in the .txt file represents one keypoint. 
- The first value is the class identifier (e.g., 0 for a person).
- The next two values are the normalized `x` and `y` center coordinates of the keypoint.
- The following two values could represent the width and height of the bounding box around the keypoint, though they might be less relevant for keypoints and more for bounding boxes. 
- The last value represents the visibility of the keypoint (e.g., `1` for visible).

**Note:** For pose detection tasks specifically, you might focus primarily on the `x` and `y` coordinates. The structure could vary depending on how you plan to use this data for training.

To train a model using such annotations, the YOLOv8 architecture itself, including `yolov8n-pose`, requires modifications or an additional training setup to interpret these keypoints for activity recognition (like detecting a fight). Training directly with `model.train(data='custom-pose.yaml', epochs=100, imgsz=640)` works if your `custom-pose.yaml` points to a correctly structured dataset and annotations for object detection or segmentation, but for action or activity recognition based on keypoints, you might need a more tailored approach or additional model architecture on top of the detected keypoints to interpret sequences of movements as specific activities.

If you’re diving into pose-based activity recognition, consider exploring models optimized for sequential data analysis, like RNNs or LSTMs, to analyze the keypoints over time and deduce activities. The extracted keypoints through YOLOv8-pose can serve as valuable input for these models. 🚀
